quickfixj-core/src/test/java/quickfix/SessionSettingsTest.java

```diff
@@ -313,6 +324,57 @@ public void testVariableInterpolationWithProps() throws Exception {
                 settings.getString("VariableTest"));
     }
 
+    @Test
+    public void testVariableInterpolationWithCustomPropsForSessionIdFromInputStream() throws Exception {
+        // GIVEN
+        System.setProperty("test.2", "BAR");
+        final Properties properties = new Properties(System.getProperties());
+        properties.setProperty("test.1", "FOO");
+
+        String settingsString = getDefaultSettingString();
+        settingsString += "\n";
+        settingsString += "[SESSION]\n";
+        settingsString += "BeginString=FIX.4.2\n";
+        settingsString += "TargetCompID=CLIENT3_${test.1}_${test.2}\n";
+        settingsString += "DataDictionary=../spec/FIX42.xml\n";
+
+        // WHEN
+        final SessionSettings settings = createSettingsFromString(settingsString, properties);
+
+        // THEN
+        Iterator<SessionID> sessionIDIterator = settings.sectionIterator();
+        while (sessionIDIterator.hasNext()) {
```

Comment:
This method is both hard to follow and repeated.

I prefer

```java
private SessionID findSessionID(SessionSettings settings, String targetCompID) {
     while (sessionIDIterator.hasNext()) {
          if (sessionID.getTargetCompID().startsWith(sessionID)) {
                  return sessionID;
          }
     }
     return null;
}
```
then

```java
SessionID sessionID = findSessionID(settings, "CLIENT3");
assertNotNull("Settings for CLIENT3 not found", sessionID);
assertEquals("Wrong TargetCompID", "CLIENT3_FOO_BAR", sessionID.getTargetCompID());
```

This abstracts the nasty Iterator iteration out the test and makes the test closer to the arrange, act, assert model - which is what your comments are really saying.

Reply:
Fixed